File: php/src/Models/SecureMessage.php
<?php

declare(strict_types=1);

namespace OtpCrypto\Models;

use OtpCrypto\Crypto\RandNonce;
use OtpCrypto\Crypto\Utils;
use OtpCrypto\Crypto\InvalidMessageException;

/**
 * SecureMessage – Wire model (headers/body)
 * -----------------------------------------
 * Represents the protocol wire format that travels alongside HTTP:
 *
 *   Headers:
 *     "v": 1                     // protocol version
 *     "w": <int>                 // time window (floor(epoch/30))
 *     "n": "<b64_nonce>"         // 8-byte random nonce (Base64)
 *     "c": "<b64_ciphertext>"    // AES-256-CBC ciphertext (Base64)
 *
 *   Body:
 *     "<b64_tag>"                // 32-byte HMAC-SHA256 tag (Base64)
 *
 * This class:
 *  - Parses/validates wire fields (format only; no crypto verification).
 *  - Holds binary values (nonce/ciphertext/tag) after Base64 decoding.
 *  - Serializes back to wire form (Base64) for transport.
 *
 * SECURITY:
 *  - Does NOT verify MAC or decrypt; higher layers must do that.
 *  - Keeps messages generic on errors via InvalidMessageException.
 */
final class SecureMessage
{
    /** Protocol version (`v`). */
    private int $v;

    /** Time window (`w`). */
    private int $w;

    /** 8-byte nonce (binary). */
    private string $nonce;

    /** Ciphertext (binary). */
    private string $ciphertext;

    /** 32-byte HMAC-SHA256 tag (binary). */
    private string $tag;

    private function __construct(
        int $v,
        int $w,
        string $nonce,
        string $ciphertext,
        string $tag
    ) {
        $this->v = $v;
        $this->w = $w;
        $this->nonce = $nonce;
        $this->ciphertext = $ciphertext;
        $this->tag = $tag;
    }

    /**
     * Builds a SecureMessage from raw (already-decoded) parts.
     *
     * @param int    $v           Protocol version (>=1).
     * @param int    $w           Time window (>=0).
     * @param string $nonce       8-byte binary nonce.
     * @param string $ciphertext  Binary ciphertext.
     * @param string $tag         32-byte binary HMAC tag.
     * @return self
     *
     * @throws InvalidMessageException on any format error.
     *
     * HINT: Use this when the caller already decoded Base64 fields.
     */
    public static function fromParts(
        int $v,
        int $w,
        string $nonce,
        string $ciphertext,
        string $tag
    ): self {
        try {
            if ($v < 1) {
                throw new \InvalidArgumentException('protocol version must be >= 1');
            }
            if ($w < 0) {
                throw new \InvalidArgumentException('window must be non-negative');
            }
            RandNonce::validate($nonce);
            if ($ciphertext === '') {
                throw new \InvalidArgumentException('ciphertext must not be empty');
            }
            if (\strlen($tag) !== 32) {
                throw new \InvalidArgumentException('tag must be 32 bytes');
            }
            return new self($v, $w, $nonce, $ciphertext, $tag);
        } catch (\Throwable $e) {
            throw new InvalidMessageException($e);
        }
    }

    /**
     * Parses a SecureMessage from **wire** headers/body (Base64 strings).
     *
     * @param array  $headers Must contain keys: "v","w","n","c" (strings).
     * @param string $body    Base64 HMAC tag string.
     * @return self
     *
     * @throws InvalidMessageException if any required field is missing/invalid.
     *
     * HINT: This is a *format* parser only (no MAC/decrypt).
     */
    public static function fromWire(array $headers, string $body): self
    {
        try {
            // Required headers
            $vStr = $headers['v'] ?? null;
            $wStr = $headers['w'] ?? null;
            $nStr = $headers['n'] ?? null;
            $cStr = $headers['c'] ?? null;

            if ($vStr === null || $wStr === null || $nStr === null || $cStr === null) {
                throw new \InvalidArgumentException('missing required headers v/w/n/c');
            }

            // Parse ints
            if (!is_string($vStr) || !is_string($wStr)) {
                throw new \InvalidArgumentException('v and w must be strings');
            }
            $v = (int)$vStr;
            $w = (int)$wStr;
            if ((string)$v !== $vStr || $v < 1) {
                throw new \InvalidArgumentException('invalid v');
            }
            if ((string)$w !== $wStr || $w < 0) {
                throw new \InvalidArgumentException('invalid w');
            }

            if (!is_string($nStr) || !is_string($cStr) || !is_string($body)) {
                throw new \InvalidArgumentException('wire fields must be strings');
            }

            // Decode Base64 fields
            $nonce      = Utils::fromBase64Strict($nStr);
            $ciphertext = Utils::fromBase64Strict($cStr);
            $tag        = Utils::fromBase64Strict($body);

            // Enforce lengths
            RandNonce::validate($nonce);
            if ($ciphertext === '') {
                throw new \InvalidArgumentException('ciphertext must not be empty');
            }
            if (\strlen($tag) !== 32) {
                throw new \InvalidArgumentException('tag must be 32 bytes');
            }

            return new self($v, $w, $nonce, $ciphertext, $tag);
        } catch (\Throwable $e) {
            throw new InvalidMessageException($e);
        }
    }

    /**
     * Serializes this message to **wire headers** (Base64).
     *
     * @return array `["v"=>"1","w"=>"...","n"=>"<b64>","c"=>"<b64>"]`
     *
     * HINT: Attach this map to your HTTP request/response headers.
     */
    public function toWireHeaders(): array
    {
        return [
            'v' => (string)$this->v,
            'w' => (string)$this->w,
            'n' => Utils::toBase64($this->nonce),
            'c' => Utils::toBase64($this->ciphertext),
        ];
    }

    /**
     * Serializes the tag to **wire body** (Base64).
     *
     * @return string Base64 tag string.
     *
     * HINT: Put this as the HTTP body (string).
     */
    public function toWireBody(): string
    {
        return Utils::toBase64($this->tag);
    }

    /**
     * Clone with optional substitutions (re-validates variable-length fields).
     *
     * @param int|null    $v
     * @param int|null    $w
     * @param string|null $nonce
     * @param string|null $ciphertext
     * @param string|null $tag
     * @return self
     *
     * HINT: Useful in tests/tools; not typically needed in handlers.
     */
    public function copyWith(
        ?int $v = null,
        ?int $w = null,
        ?string $nonce = null,
        ?string $ciphertext = null,
        ?string $tag = null
    ): self {
        $nv = $v ?? $this->v;
        $nw = $w ?? $this->w;
        $nn = $nonce ?? $this->nonce;
        $nc = $ciphertext ?? $this->ciphertext;
        $nt = $tag ?? $this->tag;

        // Reuse fromParts for validation
        return self::fromParts($nv, $nw, $nn, $nc, $nt);
    }

    // --- Getters -------------------------------------------------------------

    /** Protocol version (`v`). */
    public function v(): int
    {
        return $this->v;
    }

    /** Time window (`w`). */
    public function w(): int
    {
        return $this->w;
    }

    /** 8-byte nonce (binary). */
    public function nonce(): string
    {
        return $this->nonce;
    }

    /** Ciphertext (binary). */
    public function ciphertext(): string
    {
        return $this->ciphertext;
    }

    /** 32-byte HMAC tag (binary). */
    public function tag(): string
    {
        return $this->tag;
    }
}

File: dart/lib/models/secure_message.dart
/// SecureMessage – Wire model for headers/body
/// -------------------------------------------
/// Represents the protocol wire format used over (or alongside) HTTP:
///   Headers (as key-value strings):
///     "version": 1                    // protocol version
///     "window": <int>                // time window (floor(epoch/30))
///     "nonce": "<b64_nonce>"        // 8-byte random nonce (Base64)
///     "ciphertext": "<b64_ciphertext>"   // AES-256-CBC ciphertext (Base64)
///   Body (as string):
///     "<b64_tag>"               // HMAC-SHA256 tag (Base64)
///
/// This type offers:
/// - Construction from raw bytes (`fromParts`)
/// - Parsing from wire headers/body (`fromWire`)
/// - Serialization back to wire headers/body (`toWireHeaders`, `toWireBody`)
///
/// SECURITY NOTES:
/// - This type **does not** verify MACs nor decrypt. It only parses/holds data.
/// - Validation here is strictly for format (presence, base64, lengths).
/// - Keep field names exactly "version","window","nonce","ciphertext" to match the protocol.
///
/// HINTS:
/// - Use with `Encryptor` (to build a SecureMessage) and `Decryptor` (to parse).
/// - Headers are returned as `Map<String,String>` ready to be attached to an
///   HTTP request by a higher-level client (this library does not send HTTP).

import 'dart:convert';
import 'dart:typed_data';

import '../otp_crypto/errors.dart';
import '../otp_crypto/rand_nonce.dart';
import '../otp_crypto/utils.dart';

class SecureMessage {
  /// Protocol version (`version`). Currently 1.
  final int version;

  /// Time window (`window`): floor(epoch / windowSeconds).
  final int window;

  /// Nonce (`nonce`) as raw 8 bytes.
  final Uint8List nonce;

  /// Ciphertext (`ciphertext`) as raw bytes.
  final Uint8List ciphertext;

  /// Tag (body) as raw 32 bytes (HMAC-SHA256).
  final Uint8List tag;

  /// Creates an immutable `SecureMessage` from already-parsed parts.
  ///
  /// HINT: Prefer `fromParts` to enforce basic length checks at creation time.
  SecureMessage._internal({
    required this.version,
    required this.window,
    required this.nonce,
    required this.ciphertext,
    required this.tag,
  });

  Map<String,dynamic> toJson() {
    return {
      'version': version,
      'window': window,
      'nonce': base64.encode(nonce),
      'ciphertext': base64.encode(ciphertext),
      'tag': base64.encode(tag),
    };
  }

  /// Builds a `SecureMessage` from *raw* components.
  ///
  /// [version] protocol version (must be >=1).
  /// [window] time window (non-negative).
  /// [nonce] 8-byte random nonce (validated).
  /// [ciphertext] AES-256-CBC ciphertext bytes.
  /// [tag] 32-byte HMAC-SHA256 tag.
  ///
  /// Throws [InvalidMessageException] if any precondition fails.
  static SecureMessage fromParts({
    required int version,
    required int window,
    required Uint8List nonce,
    required Uint8List ciphertext,
    required Uint8List tag,
  }) {
    try {
      if (version < 1) {
        throw ArgumentError.value(version, 'version', 'Protocol version must be >= 1.');
      }
      if (window < 0) {
        throw ArgumentError.value(window, 'window', 'Window must be non-negative.');
      }
      NonceGenerator.validate(nonce);
      if (ciphertext.isEmpty) {
        throw ArgumentError('ciphertext must not be empty.');
      }
      if (tag.length != 32) {
        throw ArgumentError.value(tag.length, 'tag.length', 'HMAC-SHA256 tag must be 32 bytes.');
      }

      // Store defensive copies to maintain immutability guarantees.
      return SecureMessage._internal(
        version: version,
        window: window,
        nonce: Uint8List.fromList(nonce),
        ciphertext: Uint8List.fromList(ciphertext),
        tag: Uint8List.fromList(tag),
      );
    } catch (e, st) {
      throw InvalidMessageException(cause: e, stackTrace: st);
    }
  }

  /// Parses a `SecureMessage` from **wire** headers/body.
  ///
  /// [headers] must contain keys: "version","window","nonce","ciphertext" (all strings).
  /// [body]    must be the Base64-encoded tag string.
  ///
  /// Throws [InvalidMessageException] on missing fields, bad integers/base64,
  /// nonce length != 8, or empty ciphertext.
  ///
  /// HINT: This is a *format* parser only; MAC verification is done elsewhere.
  static SecureMessage fromWire({
    required Map<String, String> headers,
    required String body,
  }) {
    try {
      // -- Required header fields -------------------------------------------
      final versionStr = headers['version'];
      final windowStr = headers['window'];
      final nonceStr = headers['nonce'];
      final ciphertextStr = headers['ciphertext'];

      if (versionStr == null || windowStr == null || nonceStr == null || ciphertextStr == null) {
        throw ArgumentError('missing required headers v/w/n/c');
      }

      // Parse integers (version, window). Reject non-integers / negatives.
      final version = int.parse(versionStr);
      final window = int.parse(windowStr);
      if (version < 1) {
        throw ArgumentError.value(version, 'version', 'Protocol version must be >= 1.');
      }
      if (window < 0) {
        throw ArgumentError.value(window, 'window', 'Window must be non-negative.');
      }

      // Decode Base64 fields.
      final nonce = Bytes.fromBase64Strict(nonceStr);
      final ciphertext = Bytes.fromBase64Strict(ciphertextStr);
      final tag = Bytes.fromBase64Strict(body);

      // Enforce lengths.
      NonceGenerator.validate(nonce);
      if (ciphertext.isEmpty) {
        throw ArgumentError('ciphertext must not be empty.');
      }
      if (tag.length != 32) {
        throw ArgumentError.value(tag.length, 'tag.length', 'HMAC-SHA256 tag must be 32 bytes.');
      }

      return SecureMessage._internal(
        version: version,
        window: window,
        nonce: nonce,
        ciphertext: ciphertext,
        tag: tag,
      );
    } catch (e, st) {
      throw InvalidMessageException(cause: e, stackTrace: st);
    }
  }

  /// Serializes this message to **wire headers** map using Base64 for binary fields.
  ///
  /// RETURNS: `{"version": "$version", "window": "$window", "nonce": "<b64>", "ciphertext": "<b64>"}`
  ///
  /// HINT: Attach this map as HTTP headers at the call site; this library does not send HTTP.
  Map<String, String> toWireHeaders() {
    return {
      'version': version.toString(),
      'window': window.toString(),
      'nonce': Bytes.toBase64(nonce),
      'ciphertext': Bytes.toBase64(ciphertext),
    };
  }

  /// Serializes this message to **wire body** string (Base64 tag).
  ///
  /// RETURNS: `"<b64_tag>"`
  ///
  /// HINT: Put this as the HTTP body (string) at the call site.
  String toWireBody() => Bytes.toBase64(tag);

  /// Convenience: clones this message with new ciphertext/tag (e.g., for re-encrypt).
  ///
  /// HINT: Rarely needed; supplied for completeness in tests/tools.
  SecureMessage copyWith({
    int? version,
    int? window,
    Uint8List? nonce,
    Uint8List? ciphertext,
    Uint8List? tag,
  }) {
    final nextNonce = nonce ?? this.nonce;
    final nextCipher = ciphertext ?? this.ciphertext;
    final nextTag = tag ?? this.tag;

    // Re-validate variable-length fields on mutation.
    NonceGenerator.validate(nextNonce);
    if (nextCipher.isEmpty) {
      throw InvalidMessageException(cause: ArgumentError('ciphertext must not be empty.'));
    }
    if (nextTag.length != 32) {
      throw InvalidMessageException(cause: ArgumentError.value(nextTag.length, 'tag.length', 'HMAC-SHA256 tag must be 32 bytes.'));
    }

    return SecureMessage._internal(
      version: version ?? this.version,
      window: window ?? this.window,
      nonce: Uint8List.fromList(nextNonce),
      ciphertext: Uint8List.fromList(nextCipher),
      tag: Uint8List.fromList(nextTag),
    );
  }
}
